pkg/kv/kvserver/replica_app_batch.go

```diff
@@ -94,39 +95,30 @@ func (b *replicaAppBatch) Stage(
 	ctx context.Context, cmdI apply.Command,
 ) (apply.CheckedCommand, error) {
 	cmd := cmdI.(*replicatedCmd)
-	if cmd.Index() == 0 {
-		return nil, errors.AssertionFailedf("processRaftCommand requires a non-zero index")
+
+	var ab appBatch
+
+	leaseIndex, rej, forcedErr, err := ab.assertAndCheckCommand(ctx, &cmd.ReplicatedCmd, &b.state, cmd.IsLocal())
```

Comment:
Yes, `ab` can only know about `ReplicatedCmd` because `ab` is standalone application. Standalone application must be entirely driven by what's in the log.